### ISSUE TITLE
Undici instrumentation marks successful MCP requests as aborted due to premature AbortController.abort() in SSE / StreamableHTTP transports

### ISSUE THREAD

**Issue by bhaveshpatel640** [modelcontextprotocol/typescript-sdk#1231]
When using the MCP TypeScript SDK (`SSEClientTransport` / `StreamableHTTPClientTransport`) with **Undici instrumentation** (e.g., OpenTelemetry), successful MCP POST requests are being reported as **aborted**.

This happens because `close()` immediately calls:

```ts
this._abortController?.abort()
```

If a POST response is still being processed, Undici receives the abort signal and marks the span as an error, even though the server returned **200 OK**. This results in spans like:

```
UND_ERR_ABORTED: The operation was aborted
```

for successful requests.

---

### **Impact**

* Successful MCP requests show up as failed in traces
* Undici logs aborted errors
* Misleading telemetry in distributed tracing
* Connections stay open until Undici’s idle timeout (~10–14 seconds)

---

### **Reproduction**

1. Enable OTel Undici instrumentation
2. Use `SSEClientTransport` or `StreamableHTTPClientTransport`
3. Send any POST request (`initialize`, `callTool`, etc.)
4. Immediately call `client.close()`
5. Observe:

   * Response is 200 OK
   * Undici reports `UND_ERR_ABORTED`
   * Span is marked as failed

---

### **Request**

Could the SDK:
1. Avoid calling `AbortController.abort()` while requests are still in-flight
2. Optionally provide a **graceful close** mode that waits for active requests to finish

This would prevent Undici from misclassifying successful MCP requests as aborted and keep telemetry accurate.

**Comment by bhaveshpatel640** [comment]
@jspahrsummers @ihrpr  

Calling `AbortController.abort()` in `close()` seems to mark in-flight 200 OK requests as aborted in Undici/OTel.  

Can we take a look at implementing a graceful close here (wait for in-flight requests before aborting)?

**Comment by bhaveshpatel640** [comment]
@felixweinberger @jspahrsummers @ihrpr

Can we prioritize this by any chance?

**Comment by aptsalt** [comment]
I'd like to take this. My approach:

Track in-flight `send()` requests in both `SSEClientTransport` and `StreamableHTTPClientTransport`. When `close()` is called, wait for pending requests to settle (with a 2s timeout to prevent hanging) before calling `AbortController.abort()`.

This means successful POST responses will complete processing before the abort signal fires, so Undici/OpenTelemetry will see the 200 OK instead of `UND_ERR_ABORTED`. Stuck requests still get cleaned up after the timeout.

PR incoming.

**Comment by travisbreaks** [comment]
I'll take this one. The issue is that `AbortController.abort()` during cleanup marks otherwise-successful SSE/Streamable HTTP requests as failed, breaking OpenTelemetry traces. I'll submit a PR with the fix.

**Comment by Necmttn** [comment]
Telemetry should distinguish close reason from request outcome. If the HTTP POST returned 200 and the response body was processed, span status should stay OK even if client close aborts a later stream or read.

A separate event or attr can record `mcp.transport.close_reason=client_close` and duration until close. Regression: successful `callTool` followed by immediate close produces a 200 OK span with no error status, while cleanup still aborts truly stuck requests after timeout.

---

_Generated with [ax](https://github.com/Necmttn/ax)._

**Comment by YashSensei** [comment]
I'd like to work on this. The root cause is clear from the issue: `close()` calls `this._abortController?.abort()` unconditionally, which fires even when a POST request is mid-flight and the server has already returned 200 OK. Undici sees the abort signal and marks the span as `UND_ERR_ABORTED`.

My plan:
1. Track in-flight requests in both `SSEClientTransport` and `StreamableHTTPClientTransport` (a counter or `Set<AbortController>` per request)
2. In `close()`, if requests are still active, either wait for them to settle (with a timeout) or only abort the *idle* connection — not the active request controllers
3. Add a regression test that sends a POST, immediately calls `close()`, and asserts the span is not marked as aborted

I'd also add an optional `gracefulClose` option (default `true`) so users who need the old immediate-abort behavior can opt out. Let me know if the team has a preference on the API shape before I start.

**Comment by YashSensei** [comment]
Took a closer look at the current Undici span reporting. I think the cleanest fix is a small interceptor that treats MCP frame boundaries as the signal of "success":

1. Watch the MCP transport stream until `mcp-RPC-end` or the equivalent JSON-RPC `result` arrives.
2. Only then mark the HTTP span status as OK.
3. If the connection drops before that frame, keep the existing error/exception path.

This avoids guessing based on HTTP 200 alone (which undici naturally reports because the TCP/TLS handshake succeeded), and it keeps the fix scope minimal — no changes to the base MCP client API.

I can put together a PR with tests if the maintainers are open to it. Any preference on whether the fix lives in the SDK's instrumentation package or as an example for users to opt into?

**Comment by K4bain** [comment]
Following up on the direction @felixweinberger laid out on #1692 (the two transport-level drains were closed because a mandatory drain adds latency to every close, and aborting in-flight requests on \close()\ is legitimate when that is what the caller asked for): I've implemented the **opt-in graceful close at the \Client.close()\ level** that was suggested there, against current \main\.

**The API:**

\\\	s
// per call
await client.close({ drainPendingRequests: true });          // default 2s cap
await client.close({ drainPendingRequests: { timeoutMs: 10_000 } });

// or once at construction (SIGINT-style shutdowns where you don't know what's in flight)
const client = new Client(info, { gracefulClose: true });
await client.close();
\\\

**Design notes:**

- \Protocol\ now tracks pending outbound request ids alongside the existing response-handler lifecycle (registered before the send, released on every exit path — response, timeout, cancel, close). \close({ drainPendingRequests })\ waits for that set to drain before tearing the transport down; requests still outstanding after the timeout fall through to the normal close path and settle with the existing connection-closed error.
- Default \close()\ behavior is unchanged. No transport-level changes — this composes with (or without) the approach in #2028, which operates one layer down.
- This addresses the undici/OTel symptom from the issue body: a 200 OK response is allowed to finish reading instead of being aborted mid-flight, so spans reflect the real outcome.
- 7 regression tests covering: unchanged default close, drain-then-close ordering, multiple in-flight requests, timeout fallback, idle immediate close, and the constructor-default/explicit-override interaction.

Branch is pushed and a PR will follow momentarily. Happy to adjust the shape (option naming, default timeout) to whatever the maintainers prefer.
